sdk/core/Azure.Core.Experimental/src/Variant/Value.cs

```diff
@@ -0,0 +1,1347 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Azure {
+    /// <summary>
+    /// Used to store primitive values without boxing, and other intstances.
+    /// </summary>
+    public readonly partial struct Value
+    {
+        private readonly Union _union;
+        private readonly object? _object;
+
+        /// <summary>
+        /// Creates instance.
+        /// </summary>
+        /// <param name="value"></param>
+        public Value(object? value)
+        {
+            _object = value;
+            _union = default;
+        }
+
+        /// <summary>
+        /// Type of the instance stored in this value.
+        /// </summary>
+        public readonly Type? Type
+        {
+            get
+            {
+                Type? type;
+                if (_object is null)
+                {
+                    type = null;
+                }
+                else if (_object is TypeFlag typeFlag)
+                {
+                    type = typeFlag.Type;
+                }
+                else
+                {
+                    type = _object.GetType();
+
+                    if (_union.UInt64 != 0 && type.IsArray)
+                    {
+                        // We have an ArraySegment
+                        if (type == typeof(byte[]))
+                        {
+                            type = typeof(ArraySegment<byte>);
+                        }
+                        else if (type == typeof(char[]))
+                        {
+                            type = typeof(ArraySegment<char>);
+                        }
+                        else
+                        {
+                            ThrowInvalidOperation();
+                        }
+                    }
+                }
+
+                return type;
+            }
+        }
+
+        [DoesNotReturn]
+        private static void ThrowInvalidCast() => throw new InvalidCastException();
+
+        [DoesNotReturn]
+        private static void ThrowArgumentNull(string paramName) => throw new ArgumentNullException(paramName);
+
+        [DoesNotReturn]
+        private static void ThrowInvalidOperation() => throw new InvalidOperationException();
+
+        #region Byte
+        /// <summary>
+        /// Stores byte in this value.
+        /// </summary>
+        /// <param name="value"></param>
+        public Value(byte value)
+        {
+            this = default;
+            _object = TypeFlags.Byte;
+            _union.Byte = value;
+        }
+
+        /// <summary>
+        /// Stores nullable byte in this value.
+        /// </summary>
+        /// <param name="value"></param>
+        public Value(byte? value)
+        {
+            this = default;
+            if (value.HasValue)
+            {
+                _object = TypeFlags.Byte;
+                _union.Byte = value.Value;
+            }
+            else
+            {
+                _object = null;
```

Comment:
I don't like the fact that the type is now null. I think we should preserve the type

Reply:
In this case, would we create a NullByte TypeFlag and assume that means the value is of type Byte and also null?

Reply:
Yeah, I think this is the way to do it. But we need to think through all the implications. @jeremykuhne, do you remember why you decided to not preserve the type for nullable values? I can think of many scenarios where it would be problematic. For example, if a table value is no-null, the caller can tell/query the corresponding column's type. But if the value happens to be null, then the type of the column cannot be determined.

Reply:
I didn't store the type for any null value. You can certainly support nulls by storing the `Type` as the `_object`, it would just require some extra branches to check for that case.

Reply:
OK. I will play with this in a subsequent PR.